File: PermutationsII.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        nums.sort()
        self.dfs(nums,[],res)
        return res
        
    def dfs(self,nums,path,res):
        if not nums:
            res.append(path)
        i = 0
        while i < len(nums):
            self.dfs(nums[:i] + nums[i+1:], path + [nums[i]],res)
            while i+1 < len(nums) and nums[i] == nums[i+1]:
                i += 1
            i += 1
        

File: permutationsII.kt
class Solution {
    fun permuteUnique(nums: IntArray): List<List<Int>> {
        val temp = mutableListOf<Int>()
        val res = mutableListOf<List<Int>>()
        val picked = hashMapOf<Int, Int>()
        for (num in nums){
            picked[num] = picked.getOrDefault(num, 0) + 1
        }
        dfs(nums, temp, picked, res)
        return res.toList()
    }

    fun dfs(nums: IntArray, temp: MutableList<Int>, picked: HashMap<Int, Int>, res: MutableList<List<Int>>) {
        if (temp.size == nums.size) {
            res.add(temp.toList())
            return
        }
        for ((num, count) in picked){
            if (count > 0){
                picked[num] = picked[num]!! - 1
                temp.add(num)
                dfs(nums, temp, picked, res)
                temp.removeAt(temp.size -1) // remove from the last position
                picked[num] = picked[num]!! + 1
            }
        }
    }
}
